### ISSUE TITLE
Windows (Fastcall) calling convention: Callee-saved XMM (FP) registers are not actually saved

### ISSUE THREAD

**Issue by hrydgard** [bytecodealliance/cranelift#1366]
We're having issues, only on Windows, with floating point variables becoming corrupt across calls into Cranelift from Wasmer.

The following function `callee_saved_gprs` in the codegen is supposed to return a list of the callee-saved registers:

https://github.com/bytecodealliance/cranelift/blob/23e9bdb2d99fb8f554793cb87f1578cc543c355f/cranelift-codegen/src/isa/x86/abi.rs#L370

However, it, nor its callers, seem to concern themselves with saving the floating point callee-saved registers. Quoth https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019#callercallee-saved-registers:

> The registers RBX, RBP, RDI, RSI, RSP, R12, R13, R14, R15, and XMM6-15 are considered nonvolatile and must be saved and restored by a function that uses them.

`callee_saved_gprs` only lists RBX, RBP, RDI, RSI, RSP, R12, R13, R14, R15, ignoring XMM6-15.

I believe this may be the cause of our troubles, it definitely looks wrong. What do you think?

**Comment by sstangl** [comment]
Windows does indeed require restoring those registers on exit if changed.

IonMonkey implementation for reference: https://searchfox.org/mozilla-central/rev/cbd110d718bc89a499d3f996af24532abbf6f8ea/js/src/jit/x64/Trampoline-x64.cpp#72

**Comment by hrydgard** [comment]
We've done more investigation, and manually saving and restoring the registers around calls into Cranelift using some inline assembly fixes our variable corruption issues. 

I don't know if this is actually a real security issue, since values are leaking from VM to host, and the host is in control of the VM's memory anyway. I guess it's possible that values could leak from host to VM if there are bugs in the register allocator, since nothing seems to wipe the XMM registers when entering the VM (though I haven't looked closely for that).

Either way, this clearly needs to be addressed!

**Comment by sstangl** [comment]
> We've done more investigation, and manually saving and restoring the XMM registers around calls into Cranelift using some inline assembly fixes our variable corruption issues.

Well yes, that's logically equivalent to fixing Cranelift's Microsoft x64 calling convention implementation. Except in the case of Cranelift ever calling out to native code, in which case that boundary would also need to know to save/restore registers.

The workaround isn't terrifying -- that's pretty much what Cranelift needs to do internally, except that Cranelift can know what XMM regs were actually allocated and therefore can skip saving/restoring unused volatile registers.

**Comment by hrydgard** [comment]
Right, though I suspect that the calling-out-to-native-code scenario is not actually an issue, at least not a correctness one, because Cranelift internally doesn't seem to be aware of the concept of callee-save FP/SIMD registers, so they would simply already be making sure that they're all "dead" when calling out.

**Comment by hrydgard** [comment]
So is anyone from Cranelift going to look at this, or should I just mash together a hacky PR that saves the registers on entrance and restores them on exit, like we did in our code? That way Cranelift can keep pretending that the Windows ABI does not have callee-save FP registers...

**Comment by sstangl** [comment]
Although I suppose we'd take a hacky-but-correct patch, the real patch shouldn't be much more work.

You'd just need to make an analogue of the `callee_saved_gprs_used()` in [fastcall_prologue_epilogue()](https://github.com/bytecodealliance/cranelift/blob/master/cranelift-codegen/src/isa/x86/abi.rs#L507). Probably get that function to return both GPRs and FPRs used, so that you don't walk the CFG twice.

`insert_common_epilogue()` would need to learn about FPRs also.

**Comment by hrydgard** [comment]
Oh, you're actually part of the project, I misunderstood from your tone before that you were just another passersby :)  I mean, this is a critical bug that can cause really tricky secondary bugs and could possibly (although not super likely) be a CVE-worthy security issue, I initially kind of expected you guys to pretty much take this and run with it.

Anyway, I don't have the time to go deep here, so a patch from us would likely just be what I said before, just save and restore all the registers on entry/exit. I agree that it's probably not that much work to fix it properly, but I don't know this code base at all - I've written JIT compilers before, things are not always as easy as they first seem...

**Comment by iximeow** [comment]
Having read the MSDN x86_64 calling convention page more than a few times I didn't remember it mentioning XMM registers being nonvolatile before... it turns out even in January of last year they weren't mentioned: so says [the page via archive.org](https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019). It got added somewhere between then and [October](https://web.archive.org/web/20191030045048/https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019). Not a huge fan of this getting a few words added towards the end of one page!!

Regardless, I'm putting together a change to preserve xmm6-xmm15.
